File: 조준장/골드5/9251.py
"""
- dp
"""
import sys

a = str(sys.stdin.readline())
b = str(sys.stdin.readline())
len_a = len(a)
len_b = len(b)
dp = [[0] * len_b for _ in range(len_a)]

for i in range(1, len_a):
    for j in range(1, len_b):
        if a[i - 1] == b[j - 1]:
            dp[i][j] = dp[i - 1][j - 1] + 1
        else:
            dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

print(dp[len_a-1][len_b-1])

File: 조준장/골드5/9251.kt
class `9251` {
    fun solve() {
        val br = System.`in`.bufferedReader()
        val bw = System.out.bufferedWriter()
        val sb = StringBuilder()

        val a = br.readLine().toString()
        val b = br.readLine().toString()

        sb.append(lcsLength(a, b))

        bw.write(sb.toString())

        br.close()
        bw.close()
    }

    fun lcsLength(a: String, b: String): Int {
        val lenA = a.length
        val lenB = b.length

        val dp = Array(lenA + 1) { IntArray(lenB + 1) }

        for (i in 1..lenA) {
            for (j in 1..lenB) {
                if (a[i - 1] == b[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1] + 1
                } else {
                    dp[i][j] = maxOf(dp[i - 1][j], dp[i][j - 1])
                }
            }
        }

        return dp[lenA][lenB]
    }

}

fun main() {
    `9251`().solve()
}
